bondpy migration to ROS2

gros-antoine opened the pull request:
As bondpy was never updated to ROS2.

- the code migration is mainly done in ca826e277987c1483ccd529dd49562fa515d8145
- the other commits are mainly converting the package to a ROS2 package

Whereas the migration to ROS2 [readme](https://github.com/ros/bond_core/blob/ros2/ros2_migration_readme.md) mentions multiple changes to the internal functioning of the C++ package, I have only mainly converted the code to work with ROS2. The only current change to the way the package works is the following:

> Node is passed to a constructor of bond class.

gros-antoine committed:
remove to build

gros-antoine committed:
not necessary for pure python pkg

gros-antoine committed:
move source code for ROS2 python pkg + translate code to ROS2

gros-antoine committed:
update for ROS2

gros-antoine committed:
add for ROS2 package

gros-antoine committed:
add basic ros2 tests

gros-antoine committed:
test for connection and disconnection

gros-antoine committed:
use function to convert ros2 duration to seconds everywhere

carmiac commented:
I would really like for this or similar to get merged and eventually released as part of a regular ros release. Is there anything I can do to help that happen?

mjcarroll commented:
Apologies, this fell off my radar, I will add it to my review list for this week.

MattpMatt commented:
Any updates on when this will be merged?

clalancette commented on bondpy/bondpy/bondpy.py:
```diff
@@ -91,10 +93,11 @@ class Bond(object):
     # \param on_broken callback that will be called when the bond is broken.
     # \param on_formed callback that will be called when the bond is formed.
-    def __init__(self, topic, id, on_broken=None, on_formed=None):
+    def __init__(self, topic, id, node, on_broken=None, on_formed=None):
```
I think it would be a bit nicer to have `node` be the first argument; that more closely matches how we do things elsewhere.

gros-antoine replied:
Changed in [4949755](https://github.com/ros/bond_core/pull/89/commits/4949755d949cb878b38efabff13963561a70a10d)

clalancette commented on bondpy/bondpy/bondpy.py:
```diff
@@ -304,7 +307,7 @@ def set_broken_callback(self, on_broken):
     #
     # \param timeout Maximum duration to wait.  If None then this call will not timeout.
-    # \return true iff the bond has been formed.
+    # \return true if the bond has been formed.
```
```suggestion
    # \return true iff the bond has been formed.
```

(`iff` means "if and only if")

gros-antoine replied:
Changed in [74805b1](https://github.com/ros/bond_core/pull/89/commits/74805b1f61bde433b7f39eaf3db5ecf52190c80f)

clalancette commented on bondpy/bondpy/bondpy.py:
```diff
@@ -313,20 +316,20 @@ def wait_until_formed(self, timeout=None):
     #
     # \param timeout Maximum duration to wait.  If None then this call will not timeout.
-    # \return true iff the bond has been broken, even if it has never been formed.
+    # \return true if the bond has been broken, even if it has never been formed.
```
```suggestion
    # \return true iff the bond has been broken, even if it has never been formed.
```

gros-antoine replied:
Changed in [74805b1](https://github.com/ros/bond_core/pull/89/commits/74805b1f61bde433b7f39eaf3db5ecf52190c80f)

clalancette commented on bondpy/bondpy/bondpy.py:
```diff
@@ -335,18 +338,18 @@ def wait_until_broken(self, timeout=None):
 
     ## \brief Indicates if the bond is broken
-    # \return true iff the bond has been broken.
+    # \return true if the bond has been broken
```
```suggestion
    # \return true iff the bond has been broken
```

gros-antoine replied:
Changed in [74805b1](https://github.com/ros/bond_core/pull/89/commits/74805b1f61bde433b7f39eaf3db5ecf52190c80f)

clalancette commented on bondpy/package.xml:
```diff
@@ -15,17 +15,20 @@
+    <build_type>ament_python</build_type>
   </export>
-</package>
+</package>
```
Please add a trailing newline.

gros-antoine replied:
Changed in [57fd471](https://github.com/ros/bond_core/pull/89/commits/57fd47196927dd08479548c6495c1883cf621561)

clalancette reviewed (changes requested):
Overall, this looks like a good port to me.

There are a few things that still need to be fixed up in this package:
1.  There are a few changes I've left inline that I think should be fixed.
2.  This needs to be rebased onto the latest, and conflicts in setup.py fixed.
3.  When running tests, there are a lot of flake8 problems that should be fixed.
4.  The copyright test should be un-skipped, and we should add the copyright headers as appropriate.

Once those things are done, I think this will be ready for merge.

gros-antoine commented:
@clalancette Thanks a lot for the review, I'll take a look at it as soon as possible next week :)

gros-antoine committed:
Merge remote-tracking branch 'upstream/ros2' into ros2-bondpy

gros-antoine committed:
add missing trailing newline

gros-antoine committed:
change if to iff in comments

gros-antoine committed:
change node argument position

gros-antoine committed:
adapt test to new node argument position

gros-antoine committed:
add 'data_files' argument for marker file

gros-antoine committed:
adapt to current source code architecture

gros-antoine committed:
add pytest

gros-antoine committed:
reformat code to pass flake8 test

gros-antoine commented:
So I've changed pretty much everything that needed to be changed except for the copyright. flake8 still flags '##' used for Doxygen but I think that's ok.

Regarding the copyright, sorry I'm not familiar with this kind of topic. Do I need to add the classic copyright text for _Open Source Robotics Foundation, Inc_ at the top of ``test_bond.py`` updated to 2023 and call it a day? If someone could guide me on this, that'd be pretty much appreciated :)

clalancette commented:
> Regarding the copyright, sorry I'm not familiar with this kind of topic. Do I need to add the classic copyright text for _Open Source Robotics Foundation, Inc_ at the top of `test_bond.py` updated to 2023 and call it a day?

Yes, that should do it.

Also, when running the tests on this I'm getting a bunch of flake8 errors.  Once we fix the copyright and the flake8 errors, we can get this in.

RichFel commented:
@clalancette @gros-antoine
Hi, this looks great, is there a reason why pull request hasn't been merged? I could help out if necessary.

clalancette commented:
> Hi, this looks great, is there a reason why pull request hasn't been merged? I could help out if necessary.

It is waiting on response to my latest review.

RichFel commented:
@clalancette What flake8 errors do you have when running tests? Right now with pycodestyle 2.8.0 I have double # errors used for Doxygen and that's it. A screenshot of the warnings would greatly help! Is pycodestyle 2.8.0 the good version for this?
![image](https://github.com/ros/bond_core/assets/56121033/bf69d7d3-8dbe-42da-8325-84092f049669)

clalancette commented:
It's been a while, but I had copyright and flake8 errors.  If you install our development tools as suggested in http://docs.ros.org/en/rolling/Installation/Alternatives/Ubuntu-Development-Setup.html, you'll see the full set.

gros-antoine committed:
add copyright

gros-antoine committed:
update to pass flake8 tests

gros-antoine commented:
Hello @clalancette @RichFel,

Sorry it's been a while, started my first dev job and kind of forgot about this PR

So I've updated the missing copyrights and the code to pass the latest flake8 tests. I had to add one copyright to ```BondSM_sm.py``` which I think was auto-generated so maybe it wasn't best practice?

What still need to be fixed:

- copyright test in ```bond.py``` fails as there is a copyright but no license in it. I don't know if I can just update it to a new copyright form as it was written in 2009 and mentions Willow Garage.
- those ```##```. I thought of replacing them with the 3 quotes docstrings but I don't know if ROS's Doxygen is setup to understand them.

If anyone could guide me, that would be greatly appreciated!

clalancette committed:
Fix linting errors.

Signed-off-by: Chris Lalancette <clalancette@gmail.com>

clalancette commented:
> * copyright test in `bond.py` fails as there is a copyright but no license in it. I don't know if I can just update it to a new copyright form as it was written in 2009 and mentions Willow Garage.

It just needs a few tweaks to make it conform to the style.  I'll push a commit shortly which does that.

> * those `##`. I thought of replacing them with the 3 quotes docstrings but I don't know if ROS's Doxygen is setup to understand them.

We don't use doxygen for python at all.  The 3 quotes docstrings would be the way to make it so that sphinx pulls it appropriately.

That said, I think we can leave that for another PR.  For now, I'll push a commit that just removes the extra `#` so it passes the linters.

clalancette committed:
Cleanup setup.py.

Make it look like the rest of the ROS 2 packages.

Signed-off-by: Chris Lalancette <clalancette@gmail.com>

clalancette committed:
Stop checking BondSM_sm.py for copyright/flake8.

It is generated code, so we should not change it and
we should exclude it from linters.

Signed-off-by: Chris Lalancette <clalancette@gmail.com>

clalancette commented:
@ros-pull-request-builder retest this please

clalancette committed:
Stop excluding BondSM_sm from flake8.

This seems to be triggering some kind of bug in flake8
itself.

Signed-off-by: Chris Lalancette <clalancette@gmail.com>

clalancette committed:
Revert "Stop excluding BondSM_sm from flake8."

This reverts commit bf428c70daab9aab0df0bc1aaf13e04c123bba67.

clalancette commented:
All right.  Something really weird that I don't understand is happening with the Rpr job.  I don't think it is directly related to this PR, so with the latest changes I'm going to go ahead and merge this.  Thanks @gros-antoine for sticking with this and getting it done.

RichFel commented:
@clalancette and @gros-antoine thank you for the work you've done. Do you have any idea when this update will be released on the ROS index? I want to use the nav2 lifecycle manager for my CPP and Python nodes, but currently, I can only use it for CPP nodes since it uses bond-core as its heartbeat mechanism.

gros-antoine commented:
@clalancette Thanks a lot for your time, help and guidance :)
